lib/scholar/metrics/similarity.ex

```diff
@@ -0,0 +1,63 @@
+defmodule Scholar.Metrics.Similarity do
+  @moduledoc """
+  Similarity metrics between 1-D tensors.
+  """
+
+  import Nx.Defn
+  import Scholar.Shared
+
+  @doc ~S"""
+  Calculates Jaccard similarity (also known as Jaccard similarity coefficient, or Jaccard index).
+
+  Jaccard similarity is a statistic used to measure similarities between two sets. Mathematically, the calculation
+  of Jaccard similarity is the ratio of set intersection over set union.
+
+  $$
+  J(A, B) = \frac{\mid A \cap B \mid}{\mid A \cup B \mid}
+  $$
+
+  ## Examples
+
+      iex> x = Nx.tensor([1.0, 5.0, 3.0, 6.7])
+      iex> y = Nx.tensor([5.0, 2.5, 3.1, 9.0])
+      iex> Scholar.Metrics.Similarity.jaccard(x, y)
+      #Nx.Tensor<
+        f32
+        0.1428571492433548
+      >
+
+      iex> x = Nx.tensor([1, 2, 3, 5, 7])
+      iex> y = Nx.tensor([1, 2, 4, 8, 9])
+      iex> Scholar.Metrics.Similarity.jaccard(x, y)
+      #Nx.Tensor<
+        f32
+        0.25
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2, 3])
+      iex> Scholar.Metrics.Similarity.jaccard(x, y)
+      ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+  """
+  defn jaccard(x, y) do
+    # We're requiring the same shape because usual use cases will have the same shape.
+    # The last axis could in theory be different on both sides.
+    assert_same_shape!(x, y)
+
+    x_size = unique_size(x)
+    y_size = unique_size(y)
+
+    union_size = unique_size(Nx.concatenate([x, y]))
+    intersection_size = x_size + y_size - union_size
+
+    intersection_size / union_size
+  end
+
+  defnp unique_size(tensor) do
+    sorted = Nx.sort(tensor)
+
+    Nx.not_equal(sorted[0..-2//1], sorted[1..-1//1])
```

Comment:
This is going to fail for tensors with length 1, we should match on the shape and for `{1}` we return `1` :)

Reply:
Also worth a test!

Reply:
To be honest I don't even know what this piece of code is doing.. 

`Nx.not_equal(sorted[0..-2//1], sorted[1..-1//1]) `??

Reply:
My point is that we are accessing index 1, which is not valid for length-1 tensor.

This code creates a binary mask meaning "is an element at this position different from its successor". For example given `[3, 2, 1]` we want to compare `[3, 2]` with their successors `[2, 1]` (the last element has no successor), so we need all-but-last and all-but-first slices of the original tensor.

Reply:
I was referring to: `0..-2//1`. this range is odd to me.

Reply:
@jonatanklosko changes added

Reply:
@tiagodavi negative indices are a way of indexing from right to left instead of left to right. -2 is the second position from the right (a.k.a second to last item). For a tensor of size 5, index -2 is the same as index 3, index -1 is the same as index 4, which is the last index.

Reply:
So 0..-2//1 is a way of saying "take all items from the first until the second to last, inclusive"

Reply:
It’s similar to list index, however with double slash and a number at the end.

Reply:
Thanks